scipy/signal/_upfirdn_apply.pyx

```diff
@@ -65,6 +65,56 @@ def _output_len(Py_ssize_t len_h,
     return need
 
 
+@cython.cdivision(True)  # faster modulo
+@cython.boundscheck(False)  # designed to stay within bounds
+@cython.wraparound(False)  # we don't use negative indexing
+def _rat(x, tol=None):
```

Comment:
This code finds the continued fraction expansion up to a tolerance so you don't end up with gigantic up and down values. I actually think this function should probably be made public, but I'm not sure where it would go. Probably `numpy` somewhere...

I played around with GCD at first, but was worried it could still lead to huge filters for really long signals (where up and down could each be large). But maybe we should use that to guarantee our output length is correct, given that the signature is something like `signal.resample(x, output_length)`?

When using `Fraction(x).limit_denominator()` I couldn't get satisfyingly stable behavior, but maybe that's just me not using it properly.